File: Поиск Компонент Связности за O(n+m).cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> comp;
void _dfs (vector<bool> &used, vector<vector<int> > &g,int v)
{
	used[v] = true;
	comp.push_back(v);
	for (int i = 0; i<g[v].size(); ++i)
    {
		int to = g[v][i];
		if (!used[to])
			_dfs (used,g,to);
	}
}

vector< vector<int> > find_comps(vector<bool> &used, vector<vector<int> > &g, int n)
{
    vector< vector<int> > ans;
    for (int i = 0;i < n;++i)
        used[i] = false;
    for (int i = 0;i < n;++i)
    {
        if (!used[i])
        {
            comp.clear();
            _dfs(used,g,i);
            vector<int> pr;
            for (int j = 0;j < comp.size();++j)
                pr.push_back(comp[j]);
            ans.push_back(pr);
        }
    }
    return ans;
}

/*
-g - массив со списками смежностей.
 used - булевский массив забитый false-ами.
 n - число вершин в графе.
-Граф хранить в виде списков смежностей.
 Если в i-ом массиве есть элемент j, значит из i можно попасть в j.
-Функция find_comps Возвращает массив с массивами вершин которые находятся в одной компоненте связности.
*/

File: Поиск Компонент Связности за O(n+m).py
comp = []
def _dfs(used:list, g:list, v:int) -> None:
    used[v] = True
    comp.append(v)
    for i in range(len(g[v])):
        to = g[v][i]
        if (not used[to]):
            _dfs(used,g,to)

def find_comps(used:list,g:list,n:int) -> list:
    ans = []
    for i in range(n):
        used[i] = False
    for i in range(n):
        if (not used[i]):
            comp.clear()
            _dfs(used,g,i)
            pr = []
            for j in range(len(comp)):
                pr.append(comp[j])
            ans.append(pr)
    return ans;

'''
-g - массив со списками смежностей.
 used - булевский массив забитый False-ами.
 n - число вершин в графе.
-Граф хранить в виде списков смежностей.
 Если в i-ом массиве есть элемент j, значит из i можно попасть в j.
-Функция find_comps Возвращает массив с массивами вершин которые находятся в одной компоненте связности.
'''